# 전처리 & 머신러닝 통합 개인과제

## Part 1. 고객 데이터 품질 개선

### 실무 시나리오

전자상거래 기업의 CRM팀은 고객별 구매 패턴을 분석하고 구매 가능성이 높은 고객을 선별하려고 합니다.  
그러나 전달받은 원본 고객 데이터에는 잘못된 자료형, 범주 표기 불일치, 결측치, 중복 데이터와 극단값이 포함되어 있어 바로 분석에 사용할 수 없습니다.

CRM팀은 **데이터 분석 담당자**로서 원본 데이터의 품질 문제를 진단하고, 이후 Part 2의 EDA와 Part 3의 구매 예측 모델링에 사용할 수 있는 분석용 데이터셋을 만들어야 합니다.

### 과제 목표

- 데이터의 구조·자료형·기초 분포를 확인하고 주요 품질 문제를 파악합니다.
- 결측치, 중복값, 이상치와 범주 표기 불일치를 분석 목적에 맞게 처리합니다.
- 조건 기반 데이터 추출과 파생변수 생성을 수행합니다.
- 전처리 결과를 검증하고 다음 Part에서 사용할 CSV 파일로 저장합니다.

### 사용 환경 및 데이터

- Python 3.X
- pandas, numpy
- `전자상거래_고객구매_원본데이터.csv`

### 제출 결과물

- Part 1 실습 노트북
- `전자상거래_고객구매_전처리완료.csv`

> 문제 1~3은 필수 문제입니다.

## 문제 1. 원본 고객 데이터 품질 진단

### 업무 상황

CRM팀에 분석 일정을 공유하기 전에 원본 데이터가 실제 분석에 사용할 수 있는 상태인지 확인해야 합니다.  
데이터를 수정하기 전에 구조와 품질 문제를 먼저 점검하고, 이후 처리해야 할 항목을 정리하세요.

### 요구사항

1. 원본 데이터를 불러오고 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
2. 컬럼별 결측치와 전체 행 기준 중복 데이터를 확인하세요.
3. 주요 범주형 컬럼의 값을 확인하여 표기 불일치 여부를 파악하세요.
4. 이후 정제가 필요한 주요 품질 문제를 간단히 정리하세요.

### 힌트

- 데이터 구조와 기초 통계량을 함께 확인하면 자료형 오류와 비정상 범위를 찾기 쉽습니다.
- 범주형 컬럼은 고유값을 확인하여 대소문자, 공백, 한글·영문 혼용 여부를 살펴볼 수 있습니다.

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

DATA_PATH = Path("전자상거래_고객구매_데이터.csv")

# TODO: 원본 데이터를 불러와 df에 저장하세요.
df = pd.read_csv(DATA_PATH)
df

# TODO: 데이터의 크기, 컬럼, 자료형과 기초 통계량을 확인하세요.
print("데이터 크기:", len(df)) 
print("컬럼:", df.columns)
print("자료형:", type(df))

# 기초통계량
pd.set_option('display.float_format', '{:.2f}'.format) # 과학적 표기법 해지
display(df.describe()) 

# TODO: 결측치, 중복 데이터와 주요 범주형 컬럼의 값을 확인하세요.
### 출력 결과가 너무 많아서 끊길 수 있습니다 ###

#df_info()

# nan값(결측치)가 하나라도 포함된 칼럼 확인
missing_data = df.isna().sum()[df.isna().sum() > 0] 
display(missing_data)

# 중복된 행 확인
duplicated_data = df[df.duplicated()] 
display(duplicated_data)

# 범주형 컬럼값
categorical_columns = df.select_dtypes(include=["object"]).columns
print(categorical_columns)  
for col in categorical_columns:
    display(col, df[col].unique()) # 범주형 컬럼의 고유값들 확인


데이터 크기: 1560
컬럼: Index(['CustomerID', 'Gender', 'Age', 'Region', 'MembershipLevel',
       'VisitCount', 'AveragePurchaseAmount', 'TotalPurchaseAmount',
       'SatisfactionScore', 'LastPurchaseDate', 'CouponUsed',
       'PreferredCategory', 'SignupDate', 'Email', 'PurchaseStatus'],
      dtype='str')
자료형: <class 'pandas.DataFrame'>


,VisitCount,TotalPurchaseAmount,SatisfactionScore,PurchaseStatus
count,1560.00,1560.00,1521.00,1560.00
mean,9.52,484140.19,3.66,0.49
std,8.15,1410878.50,0.76,0.50
min,1.00,5000.00,1.10,0.00
25%,7.00,186950.00,3.20,0.00
50%,9.00,336600.00,3.70,0.00
75%,11.00,537075.00,4.20,1.00
max,170.00,30000000.00,5.00,1.00


Gender                   26
Age                      42
Region                   31
MembershipLevel          22
AveragePurchaseAmount    45
SatisfactionScore        39
LastPurchaseDate         30
PreferredCategory        26
dtype: int64

,CustomerID,Gender,Age,Region,MembershipLevel,VisitCount,AveragePurchaseAmount,TotalPurchaseAmount,SatisfactionScore,LastPurchaseDate,CouponUsed,PreferredCategory,SignupDate,Email,PurchaseStatus
91,CUST100419,Female,20,Seoul,Basic,6,102300.0,225100.00,2.20,2026-06-11,N,Fashion,2023-08-19,customer419@example.com,0
390,CUST100375,Male,34,Daejeon,Basic,13,148300.0,609500.00,3.30,2026-04-24,Y,Fashion,2021-08-28,customer375@example.com,0
410,CUST100190,Other,44,Incheon,Silver,13,175300.0,889100.00,4.90,2026-01-27,N,Sports,2023-05-23,customer190@example.com,1
435,CUST101284,Male,45,Gyeonggi,Gold,12,143400.0,904800.00,3.00,2026-05-02,Y,Beauty,2021-10-26,customer1284@example.com,1
439,CUST101200,Female,18,Busan,Basic,9,61300.0,196000.00,3.00,2025-12-15,N,Fashion,2025-05-23,customer1200@example.com,0
563,CUST100326,Male,33,Gyeonggi,Gold,13,133100.0,918600.00,3.70,2026-04-27,Y,Sports,2024-06-04,customer326@example.com,1
586,CUST100595,Male,42,Gyeonggi,Basic,10,143400.0,728100.00,3.00,2026-04-28,N,Food,2021-08-17,customer595@example.com,0
621,CUST101404,Female,25,Gyeonggi,Gold,6,135600.0,353400.00,4.80,2025-09-30,N,Fashion,2020-10-27,customer1404@example.com,0
661,CUST100994,Female,48,Seoul,VIP,4,191200.0,204800.00,NaN,2026-05-02,Y,Home,2020-09-17,customer994@example.com,0
688,CUST100636,female,42,Gyeonggi,VIP,9,209300.0,395100.00,2.70,2026-02-16,No,Food,2025-05-27,customer636@example.com,1


Index(['CustomerID', 'Gender', 'Age', 'Region', 'MembershipLevel',
       'AveragePurchaseAmount', 'LastPurchaseDate', 'CouponUsed',
       'PreferredCategory', 'SignupDate', 'Email'],
      dtype='str')


C:\Users\pulmo\AppData\Local\Temp\ipykernel_11080\2666063866.py:34: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include=["object"]).columns


'CustomerID'

<StringArray>
['CUST101486', 'CUST100113', 'CUST100168', 'CUST101212', 'CUST100312',
 'CUST101049', 'CUST100600', 'CUST100050', 'CUST101364', 'CUST100518',
 ...
 'CUST100416', 'CUST100949', 'CUST100258', 'CUST100392', 'CUST100809',
 'CUST101044', 'CUST100737', 'CUST100506', 'CUST100164', 'CUST101260']
Length: 1500, dtype: str

'Gender'

<StringArray>
[  'Female',     'Male',   'female',    'Other', ' FEMALE ',     'male',
        'F',        nan,   ' MALE ',        'M']
Length: 10, dtype: str

'Age'

<StringArray>
[     '40',      '39',     '105',      '44',      '37',      '45',      '47',
      '26',      '56',      '42',      '53',      '23',      '34',      '46',
      '52',      '31',      '29',      '43',      '50',      '18',      '25',
      '57',      '20',      '54',      '22',      '51',      '21',      '41',
      '28',      '38',      '49',      '27',       nan,      '62',      '24',
      '30',      '36',      '33',      '63',      '35',      '32',      '60',
      '48',      '61',      '58',     '29세',     '33세',     '52세',      '55',
      '59',    'none',      '19',     '44세',      '64',      '72',      '79',
     '120',      '76',      '65',      '66',     '27세',     '55세',     '43세',
       '-',     '112',      '70',     '40세',      '69',     '66세',     '130',
      '68',       '?',      '99',     '미입력',     '101',     '36세',  'thirty',
      '67', 'unknown']
Length: 79, dtype: str

'Region'

<StringArray>
[         nan, ' GYEONGGI ',    'Gwangju',      'Daegu',   'Gyeonggi',
    'Daejeon',      'Busan',      'Seoul',    'Incheon',      'seoul',
    ' Seoul ',   'gyeonggi',      'busan',    ' Busan ']
Length: 14, dtype: str

'MembershipLevel'

<StringArray>
[ 'silver',   'Basic',  'Silver',     'VIP',    'Gold',    'gold',       nan,
 ' Basic ',  'SILVER',  ' Gold ',   'basic']
Length: 11, dtype: str

'AveragePurchaseAmount'

<StringArray>
[ '75000.0',  '74400.0', '185800.0', '153900.0', '123700.0',  '69300.0',
 '272600.0',  '43900.0', '210800.0',  '81700.0',
 ...
  '71800.0',  '63600.0', '278900.0',  '65200.0', '199200.0', '138700.0',
 '172700.0', '134600.0', '113100.0',  '73500.0']
Length: 1023, dtype: str

'LastPurchaseDate'

<StringArray>
['2026-02-07', '2026-04-22', '2026-06-06', '2026-05-07', '2026-04-16',
          nan, '2026-01-31', '2026-01-21', '2026-06-12', '2026-03-19',
 ...
 '2025-11-23', '2025-11-13', '2026.04.02', '2025-11-03', '2026.04.18',
 '2026.05.27', '2026-02-23', '2026/04/05', '2025-10-27', '2025-11-02']
Length: 252, dtype: str

'CouponUsed'

<StringArray>
['N', ' n ', 'Y', 'No', 'Yes', ' y ']
Length: 6, dtype: str

'PreferredCategory'

<StringArray>
[       'Beauty',          'Home',       'Fashion',          'Food',
   'Electronics',        'beauty',        'Sports',      ' Beauty ',
       'fashion',       'FASHION', ' Electronics ',             nan,
   'electronics']
Length: 13, dtype: str

'SignupDate'

<StringArray>
['2022-12-03', '2021-07-16', '2022-02-15', '2021-10-27', '2025-11-08',
 '2024-01-05', '2025-01-28', '2021-07-01', '2024-05-03', '2023-10-13',
 ...
 '2025-02-20', '2022-05-05', '2023-01-25', '2021-01-05', '2023-11-11',
 '2024-02-26', '2020-07-06', '2025-02-05', '2021-09-24', '2020-10-04']
Length: 1065, dtype: str

'Email'

<StringArray>
['customer1486@example.com',  'customer113@example.com',
  'customer168@example.com', 'customer1212@example.com',
  'customer312@example.com', 'customer1049@example.com',
  'customer600@example.com',   'customer50@example.com',
 'customer1364@example.com',  'customer518@example.com',
 ...
  'customer416@example.com',  'customer949@example.com',
  'customer258@example.com',  'customer392@example.com',
  'customer809@example.com', 'customer1044@example.com',
  'customer737@example.com',  'customer506@example.com',
  'customer164@example.com', 'customer1260@example.com']
Length: 1500, dtype: str

### 품질 진단 결과

- 자료형 정리가 필요한 컬럼: 
1. Age : 수치형으로 변환 필요(object에서 int64)

2. AveragePurchaseAmount: 수치형으로 변환 필요(object에서 float64)

3. LastPurchaseDate, SignupDate: 날짜형으로 변환 필요(object에서 datetime64)

4. CouponUsed: 범주형 또는 수치형으로 변환필요(object에서 bool(True/False), 혹은 int64(1/0)으로 변환하는게 더 좋을 것 같습니다)

- 표기 통일이 필요한 컬럼:
1. Gender: 대소문자 섞임, 줄임말('F', 'M'), 텍스트 앞뒤로 공백(' FEMALE ')이 혼재함. 표기통일 필요.

2. Age: 숫자로 표기통일 필요. 단위사용('29세'), 영문텍스트('thirty') 등이 보임.

3. Region: 대소문자 섞임, 텍스트 앞뒤로 공백(' GYEONGGI ', ' Seoul ')있음. 표기통일 필요.

4. MembershipLevel: 대소문자 섞임, 텍스트 앞뒤 공백 통일 필요.

5. LastPurchaseDate: 구분자(0000-00-00 및 0000/00/00)가 혼용되고 있음. 통일 필요.

6. CouponUsed: 대소문자 및 앞뒤공백 통일 필요.

7. PreferredCategory: 대소문자, 앞뒤공백 통일 필요.

- 결측치가 있는 컬럼:

1. nan이 포함된 컬럼들

Gender, Age, Region, MembershipLevel, AveragePurchaseAmount, SatisfactionScore, LastPurchaseDate, PreferredCategory

2. nan이 아닌 형식의 결측치가 포함된 칼럼들

Age: 'none', '-', '?', 'unknown', '미입력' 값들이 포함됨.

- 중복 데이터 확인 결과:

전체 행 기준 중복된 행이 60개(총 행개수: 1560개)가 있음.

- 이상치 확인이 필요한 컬럼:

Age 칼럼에서 112, 120와 같은 고령의 값들이 확인됨. 이상치인지 확인이 필요함.
또한, VisitCount와 TotalPurchaseAmount의 표준편차가 매우 큰데, VisitCount의 경우 max: 170로 다른값들에 비해 많이 튀는 편이고
TotalPurchaseAmount의 경우 고액의 구매값이 이상치로 잡힐 가능성이 있음. 
정리하자면, Age, VisitCount, TotalPurchaseAmount 칼럼의 이상치 확인이 필요하다고 생각합니다.

- 이후 처리할 주요 항목:

1. 대소문자 통일 및 텍스트 앞뒤 공백제거

2. Age컬럼의 결측성 특수문자 및 텍스트 변환

3. 자료형 정리(데이터 타입 변환)

4. 결측치 제거 혹은 보간(필요시)

5. 중복되는행 제거

6. 이상치 정제(필요시)



## 문제 2. 분석 가능한 고객 데이터로 정제

### 업무 상황

품질 진단 결과를 바탕으로 고객 데이터를 분석 가능한 상태로 정리해야 합니다.  
처리 과정에서 고객의 실제 구매 행동 정보가 불필요하게 손실되지 않도록 데이터 특성을 고려하세요.

### 요구사항

1. 원본 데이터를 보존한 상태에서 정제용 데이터를 생성하세요.
2. 분석에 맞지 않는 수치형·날짜형 자료형과 범주 표기를 정리하세요.
3. 결측치와 전체 행 기준 중복 데이터를 처리하세요.
4. 주요 수치형 컬럼의 이상치를 탐지하고 적절한 방법으로 처리하세요.
5. 처리 전후의 결측치, 중복값과 이상치 상태를 확인하세요.

### 힌트

- 변환할 수 없는 문자열은 결측치로 바꾼 뒤 일관되게 처리할 수 있습니다.
- 이상치는 무조건 삭제하기보다 실제 우수 고객의 행동일 가능성도 고려하세요.
- IQR은 이상치 후보를 확인하는 대표적인 방법입니다.

In [2]:
# TODO: 원본을 보존하고 정제용 데이터 df_clean을 생성하세요.
df_clean = df.copy()
df_clean

# TODO: 수치형·날짜형 자료형과 범주 표기를 정리하세요.
## 데이터 전처리
# 대소문자 표기통일 및 텍스트 앞뒤 공백 제거
target_cols = ['Gender', 'Region', 'MembershipLevel', 'CouponUsed', 'PreferredCategory']

for col in target_cols:
  df_clean[col] = df_clean[col].astype(str).str.strip().str.title()  # str.strip(): 앞뒤공백제거 / str.title(): 앞대문자, 뒤소문자 

# 각 컬럼의 고유값 확인
'''for col in target_cols:
  print(f"=== {col} ===")
  print(df_clean[col].unique())
  print()
'''

# 약어 없애기
# Gender 칼럼 
df_clean['Gender'] = df_clean['Gender'].replace({'F': 'Female', 'M': 'Male'}) # Ohter는 일단 남겨둠
df_clean['Gender'].unique()
# CouponUsed 칼럼
df_clean['CouponUsed'] = df_clean['CouponUsed'].replace({'Y': 'Yes', 'N': 'No'})
df_clean['CouponUsed'].unique()

# Age 칼럼 별개로 처리
# '세' 제거
df_clean['Age'] = df_clean['Age'].astype(str).str.replace('세', '').str.strip()
# 'thirty'를 '30'으로 변경
df_clean['Age'] = df_clean['Age'].replace('thirty', '30')

## 자료형 정리
# 수치형으로 변환
df_clean['Age'] = pd.to_numeric(df_clean['Age'], errors='coerce')  # 나머지 결측치 문자열은 nan처리
df_clean['AveragePurchaseAmount'] = pd.to_numeric(df_clean['AveragePurchaseAmount'], errors='coerce') # nan값 남겨둠
#  CouponUsed 수치형으로 변환할지 고려해보기 #

# 날짜형으로 변환
df_clean['LastPurchaseDate'] = pd.to_datetime(df_clean['LastPurchaseDate'], errors='coerce')
df_clean['SignupDate'] = pd.to_datetime(df_clean['SignupDate'], errors='coerce')

## 결과
# 자료형 확인
df_clean.info()

# 고유값 확인
# 모든 컬럼의 고유값 
#for col in df_clean.columns:
#    print(f"[{col}] : {df_clean[col].unique()}")


<class 'pandas.DataFrame'>
RangeIndex: 1560 entries, 0 to 1559
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CustomerID             1560 non-null   str           
 1   Gender                 1534 non-null   str           
 2   Age                    1513 non-null   float64       
 3   Region                 1529 non-null   str           
 4   MembershipLevel        1538 non-null   str           
 5   VisitCount             1560 non-null   int64         
 6   AveragePurchaseAmount  1501 non-null   float64       
 7   TotalPurchaseAmount    1560 non-null   float64       
 8   SatisfactionScore      1521 non-null   float64       
 9   LastPurchaseDate       1495 non-null   datetime64[us]
 10  CouponUsed             1560 non-null   str           
 11  PreferredCategory      1534 non-null   str           
 12  SignupDate             1545 non-null   datetime64[us]
 13  Email         

In [3]:
# TODO: 결측치와 전체 행 기준 중복 데이터를 처리하세요.
## 결측치 처리

## 범주형 칼럼
# 범주형 칼럼
categorical_cols = [
    "Gender",
    "Region",
    "MembershipLevel",
    "PreferredCategory"
]

# 전처리중 Nan으로 변경된 문자열 -> 다시 NaN값으로 복원
df_clean[categorical_cols] = df_clean[categorical_cols].replace(
    ["Nan"],np.nan
)

# 결측치를 Unknown으로 처리
df_clean[categorical_cols] = df_clean[categorical_cols].fillna("Unknown")


## 수치형 칼럼
# Age 칼럼: 결측치를 중앙값(median)으로 채운후 int로 변환
median_age = df_clean['Age'].median()
df_clean['Age'] = df_clean['Age'].fillna(median_age).astype(int)  # median: 39 == mode: 39

# AveragePurchaseAmount 칼럼: 결측치를 중앙값으로 채움
df_clean["AveragePurchaseAmount"] = df_clean["AveragePurchaseAmount"].fillna(
    df_clean["AveragePurchaseAmount"].median()   # median: 117500.0
)

# SatisfactionScore 칼럼: 결측치를 최빈값(mode)로 채움
df_clean["SatisfactionScore"] = df_clean["SatisfactionScore"].fillna(
    df_clean["SatisfactionScore"].median()  # mode = 3.6 (min: 1.1 / max: 10)
)

## 날짜형 칼럼 -> 이건 가짜 날짜를 생성하는 것보다는 결측치 일단 놔두기로 함.
# LastPurchaseDate 
df_clean["LastPurchaseDate"].isna().sum() # 결측치: 65개

# SignupDate       
df_clean["SignupDate"].isna().sum()  # 결측치: 15개


## 중복치 제거
# 전체 행 기준 중복치는 제거
df_clean = df_clean.drop_duplicates()

# TODO: 처리 전후 결과를 확인하세요.
print("===== 처리 전 =====")
print("행 수:", len(df))
print("결측치 수:", df.isna().sum().sum())
print("중복 행 수:", df.duplicated().sum())

display(df.head())


print("\n===== 처리 후 =====")
print("행 수:", len(df_clean))
print("결측치 수:", df_clean.isna().sum().sum())  # 날짜형 결측치 수( 64 + 15 = 79개) 남아있음
print("중복 행 수:", df_clean.duplicated().sum())

display(df_clean.head())

===== 처리 전 =====
행 수: 1560
결측치 수: 261
중복 행 수: 60


,CustomerID,Gender,Age,Region,MembershipLevel,VisitCount,AveragePurchaseAmount,TotalPurchaseAmount,SatisfactionScore,LastPurchaseDate,CouponUsed,PreferredCategory,SignupDate,Email,PurchaseStatus
0,CUST101486,Female,40,NaN,silver,5,75000.0,139800.00,3.70,2026-02-07,N,Beauty,2022-12-03,customer1486@example.com,0
1,CUST100113,Male,39,GYEONGGI,Basic,7,74400.0,159200.00,4.80,2026-04-22,N,Home,2021-07-16,customer113@example.com,1
2,CUST100168,Male,105,Gwangju,Silver,12,185800.0,1055800.00,4.50,2026-06-06,N,Beauty,2022-02-15,customer168@example.com,1
3,CUST101212,Male,44,Daegu,VIP,7,153900.0,413200.00,4.50,2026-05-07,N,Home,2021-10-27,customer1212@example.com,1
4,CUST100312,Male,37,Gyeonggi,Basic,14,123700.0,576000.00,4.00,2026-04-16,n,Home,2025-11-08,customer312@example.com,1



===== 처리 후 =====
행 수: 1500
결측치 수: 79
중복 행 수: 0


,CustomerID,Gender,Age,Region,MembershipLevel,VisitCount,AveragePurchaseAmount,TotalPurchaseAmount,SatisfactionScore,LastPurchaseDate,CouponUsed,PreferredCategory,SignupDate,Email,PurchaseStatus
0,CUST101486,Female,40,Unknown,Silver,5,75000.00,139800.00,3.70,2026-02-07,No,Beauty,2022-12-03,customer1486@example.com,0
1,CUST100113,Male,39,Gyeonggi,Basic,7,74400.00,159200.00,4.80,2026-04-22,No,Home,2021-07-16,customer113@example.com,1
2,CUST100168,Male,105,Gwangju,Silver,12,185800.00,1055800.00,4.50,2026-06-06,No,Beauty,2022-02-15,customer168@example.com,1
3,CUST101212,Male,44,Daegu,Vip,7,153900.00,413200.00,4.50,2026-05-07,No,Home,2021-10-27,customer1212@example.com,1
4,CUST100312,Male,37,Gyeonggi,Basic,14,123700.00,576000.00,4.00,2026-04-16,No,Home,2025-11-08,customer312@example.com,1


In [4]:
def get_iqr_bounds(series):
    # TODO: Q1, Q3, IQR, 하한과 상한을 반환하세요.
    # Q1, Q3 계산
    Q1 = series.quantile(0.25)
    Q3 = series.quantile(0.75)

    # IQR 계산
    IQR = Q3 - Q1

    # 이상치 하한과 상한
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    return Q1, Q3, IQR, lower_bound, upper_bound
    

# TODO: 주요 수치형 컬럼의 이상치를 탐지하고 처리하세요.

# 이상치 탐지
numeric_cols = ["Age", "VisitCount", "TotalPurchaseAmount", "AveragePurchaseAmount"]

print("=== 처리 전 이상치 확인 ===")

for col in numeric_cols:
    Q1, Q3, IQR, lower_bound, upper_bound = get_iqr_bounds(df_clean[col])

    # 이상치 탐지
    outliers = df_clean[
        (df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)
    ]
   
    print(f"{col}")
    print(f"Q1: {Q1}, Q3: {Q3}, IQR: {IQR}")
    print(f"하한: {lower_bound}, 상한: {upper_bound}")
    print(f"이상치 개수: {len(outliers)}")
    print()

# 컬럼별로 음수값이 있는지 확인
print("=== 음수 값 확인 ===")

for col in numeric_cols:
    print(f"{col}: 음수 {((df_clean[col] < 0).sum())}개")

## 수치형 컬럼 4개 모두 꼭 이상치 처리가 필요한지 잘 모르겠음(나이, 방문횟수, 총구매금액, 평균구매금액).
## 고령이 있을 수도 있고, 횟수가 실제로 많을 수도 있고, 실제로 고액구매가 이뤄져서 평균구매 금액 또한 높아졌을 수도 있다.
## 따라서 아래의 이상치 처리(iqr 경계값으로 outlier들을 대체하는 방법)를 시행하지 않기로 결정함.

'''
# 이상치 처리: IQR 범위 밖의 값을 경계값으로 대체 
for col in numeric_cols:
    Q1, Q3, IQR, lower_bound, upper_bound = get_iqr_bounds(df_clean[col])

    df_clean[col] = df_clean[col].clip(
        lower=lower_bound,
        upper=upper_bound
    )


# 처리 전후 이상치 상태 확인
print("=== 처리 후 이상치 확인 ===")

for col in numeric_cols:
    Q1, Q3, IQR, lower_bound, upper_bound = get_iqr_bounds(df_clean[col])

    outliers = df_clean[
        (df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)
    ]

    print(f"{col}: 이상치 개수 = {len(outliers)}")

'''


=== 처리 전 이상치 확인 ===
Age
Q1: 32.0, Q3: 47.0, IQR: 15.0
하한: 9.5, 상한: 69.5
이상치 개수: 11

VisitCount
Q1: 7.0, Q3: 11.0, IQR: 4.0
하한: 1.0, 상한: 17.0
이상치 개수: 15

TotalPurchaseAmount
Q1: 186425.0, Q3: 536025.0, IQR: 349600.0
하한: -337975.0, 상한: 1060425.0
이상치 개수: 53

AveragePurchaseAmount
Q1: 83975.0, Q3: 149750.0, IQR: 65775.0
하한: -14687.5, 상한: 248412.5
이상치 개수: 24

=== 음수 값 확인 ===
Age: 음수 0개
VisitCount: 음수 0개
TotalPurchaseAmount: 음수 0개
AveragePurchaseAmount: 음수 0개


'\n# 이상치 처리: IQR 범위 밖의 값을 경계값으로 대체 \nfor col in numeric_cols:\n    Q1, Q3, IQR, lower_bound, upper_bound = get_iqr_bounds(df_clean[col])\n\n    df_clean[col] = df_clean[col].clip(\n        lower=lower_bound,\n        upper=upper_bound\n    )\n\n\n# 처리 전후 이상치 상태 확인\nprint("=== 처리 후 이상치 확인 ===")\n\nfor col in numeric_cols:\n    Q1, Q3, IQR, lower_bound, upper_bound = get_iqr_bounds(df_clean[col])\n\n    outliers = df_clean[\n        (df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)\n    ]\n\n    print(f"{col}: 이상치 개수 = {len(outliers)}")\n\n'

## 문제 3. 분석용 변수 생성 및 다음 단계 데이터 준비

### 업무 상황

정제된 데이터를 CRM팀의 고객군 분석과 구매 예측 업무에 활용하려면, 원본 컬럼만으로 확인하기 어려운 고객 특성을 분석용 변수로 표현해야 합니다.  
필요한 고객을 조건에 따라 추출하고, 이후 Part 2와 Part 3에서 활용할 파생변수를 만든 뒤 최종 데이터를 저장하세요.

### 요구사항

1. 정제된 데이터에서 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.
2. 고객 분석에 활용할 파생변수를 **2개 이상** 생성하세요.
3. 생성한 파생변수의 값과 분포가 적절한지 확인하세요.
4. 최종 데이터의 결측치, 중복값, 고객 식별자와 타깃 값을 검증하세요.
5. 전처리 완료 데이터를 `전자상거래_고객구매_전처리완료.csv`로 저장하세요.

### 힌트

- 연령대, 구매 수준, 최근 구매 여부, 우수 고객 여부 등을 파생변수 후보로 고려할 수 있습니다.
- 조건 추출 결과는 별도 DataFrame으로 확인해도 되며, 최종 데이터 전체를 삭제할 필요는 없습니다.

- 고객 연령을 구간화한 `AgeGroup`을 생성하세요.
- 총구매금액을 기준으로 `PurchaseGrade`를 생성하세요.

In [5]:
REFERENCE_DATE = pd.Timestamp("2026-06-01")

# TODO: 업무적으로 의미 있는 조건을 설정하여 고객 데이터를 추출하세요.
# 최근 구매 고객 데이터(2025-06-01 에서 2026-06-01)
recent_customer = df_clean[
    (df_clean["LastPurchaseDate"].notna()) & # 최종구매일 컬럼에서 nan이 없는 값들만 고름
    (df_clean["LastPurchaseDate"] >= REFERENCE_DATE - pd.Timedelta(days=365)) # reference_date 기준으로 1년내를 추출
].copy()

print("=== 최근 구매 고객 데이터 추출 결과 ===")
print(recent_customer.head())

# TODO: 분석용 파생변수를 2개 이상 생성하세요.
# 파생변수1: 연령대별 그룹
df_clean["AgeGroup"] = pd.cut(
    df_clean["Age"],
    bins=[0, 19, 29, 39, 49, 59, float("inf")],
    labels=["10대 이하", "20대", "30대", "40대", "50대", "60대 이상"]
)

# 파생변수2: 총구매금액 기준 구매수준에 따른 그룹
df_clean["PurchaseGrade"] = pd.qcut(
    df_clean["TotalPurchaseAmount"],
    q=4,
    labels=["Low", "Medium", "High", "VIP"],
    duplicates="drop"
)

# TODO: 파생변수의 값과 분포를 확인하세요.
# 3. 파생변수 값과 분포 확인

print("=== AgeGroup 값과 분포 ===")
print(df_clean["AgeGroup"].value_counts().sort_index())

print("\n=== PurchaseGrade 값과 분포 ===")
print(df_clean["PurchaseGrade"].value_counts().sort_index())



=== 최근 구매 고객 데이터 추출 결과 ===
   CustomerID  Gender  Age    Region MembershipLevel  VisitCount  \
0  CUST101486  Female   40   Unknown          Silver           5   
1  CUST100113    Male   39  Gyeonggi           Basic           7   
2  CUST100168    Male  105   Gwangju          Silver          12   
3  CUST101212    Male   44     Daegu             Vip           7   
4  CUST100312    Male   37  Gyeonggi           Basic          14   

   AveragePurchaseAmount  TotalPurchaseAmount  SatisfactionScore  \
0               75000.00            139800.00               3.70   
1               74400.00            159200.00               4.80   
2              185800.00           1055800.00               4.50   
3              153900.00            413200.00               4.50   
4              123700.00            576000.00               4.00   

  LastPurchaseDate CouponUsed PreferredCategory SignupDate  \
0       2026-02-07         No            Beauty 2022-12-03   
1       2026-04-22         No  

In [6]:
OUTPUT_PATH = Path("전자상거래_고객구매_전처리완료.csv")

# TODO: 최종 데이터 품질을 검증하세요.
print("\n=== 최종 데이터 결측치 확인 ===")
print(df_clean.isna().sum())   # 날짜형 데이터(nan값 놔둠)를 제외하고는 결측치 없음

print("\n=== 최종 데이터 중복값 확인 ===")
print("중복 행 개수:", df_clean.duplicated().sum())

print("\n=== 고객 식별자 검증 ===")
print("전체 행 수:", len(df_clean))
print("고객 식별자 고유값 수:", df_clean["CustomerID"].nunique()) # 고객별로 각기 다르게 가지고 있음
print("고객 식별자 예시:")
print(df_clean["CustomerID"].head())

print("\n=== 타깃 값 검증 ===")
print("타깃 결측치 개수:", df_clean["PurchaseStatus"].isna().sum())
print("타깃 고유값:", df_clean["PurchaseStatus"].unique())
print("타깃 분포:")
print(df_clean["PurchaseStatus"].value_counts())


# TODO: 전처리 완료 데이터를 CSV 파일로 저장하세요.
df_clean.to_csv(
    OUTPUT_PATH,
    index=False,
    encoding="utf-8-sig"
)

print("\n저장 완료:", OUTPUT_PATH)


=== 최종 데이터 결측치 확인 ===
CustomerID                0
Gender                    0
Age                       0
Region                    0
MembershipLevel           0
VisitCount                0
AveragePurchaseAmount     0
TotalPurchaseAmount       0
SatisfactionScore         0
LastPurchaseDate         64
CouponUsed                0
PreferredCategory         0
SignupDate               15
Email                     0
PurchaseStatus            0
AgeGroup                  0
PurchaseGrade             0
dtype: int64

=== 최종 데이터 중복값 확인 ===
중복 행 개수: 0

=== 고객 식별자 검증 ===
전체 행 수: 1500
고객 식별자 고유값 수: 1500
고객 식별자 예시:
0    CUST101486
1    CUST100113
2    CUST100168
3    CUST101212
4    CUST100312
Name: CustomerID, dtype: str

=== 타깃 값 검증 ===
타깃 결측치 개수: 0
타깃 고유값: [0 1]
타깃 분포:
PurchaseStatus
0    767
1    733
Name: count, dtype: int64

저장 완료: 전자상거래_고객구매_전처리완료.csv


In [7]:
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1500 entries, 0 to 1558
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   CustomerID             1500 non-null   str           
 1   Gender                 1500 non-null   str           
 2   Age                    1500 non-null   int64         
 3   Region                 1500 non-null   str           
 4   MembershipLevel        1500 non-null   str           
 5   VisitCount             1500 non-null   int64         
 6   AveragePurchaseAmount  1500 non-null   float64       
 7   TotalPurchaseAmount    1500 non-null   float64       
 8   SatisfactionScore      1500 non-null   float64       
 9   LastPurchaseDate       1436 non-null   datetime64[us]
 10  CouponUsed             1500 non-null   str           
 11  PreferredCategory      1500 non-null   str           
 12  SignupDate             1485 non-null   datetime64[us]
 13  Email              

# 문제 4. 데이터 정제 및 파생변수 설계 근거 설명

## 업무 상황

CRM팀은 전처리 결과가 단순히 실행되는 것뿐 아니라,
적용한 처리 기준과 파생변수가 실제 분석 목적에 적합한지 확인하려고 합니다.

문제 1~3에서 수행한 결과를 바탕으로 다음 내용을 설명하세요.

## 요구사항

### 4-1. 데이터 품질 문제의 영향과 한계

1. 문제 1에서 확인한 주요 품질 문제를 한 가지 이상 선택하세요.
2. 해당 문제가 이후 EDA 또는 모델링 결과에 미칠 수 있는 영향을 설명하세요.
3. 현재 데이터만으로 판단하기 어려운 점이나 추가 확인이 필요한 조건을 작성하세요.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

1. 적용한 결측치 처리 방법을 한 가지 이상 제시하세요.
2. 해당 방법을 선택한 이유를 데이터 특성과 연결하여 작성하세요.
3. 적용한 이상치 처리 방법과 처리 시 주의할 점을 설명하세요.

### 4-3. 파생변수의 활용 의도와 한계

1. `파생변수1`과 `파생변수2`의 생성 기준을 설명하세요.
2. 각 파생변수가 Part 2 고객 특성 분석에서 어떻게 활용될 수 있는지 작성하세요.
3. 파생변수 사용 시 발생할 수 있는 정보 손실 또는 해석상의 한계를 설명하세요.

## 문제 4 결과 작성

### 4-1. 데이터 품질 문제의 영향과 한계

- 선택한 품질 문제: 

최종적으로 날짜형 변수(LastPurchaseDate, SignupDate)의 결측치와 수치형 변수들의 이상치가 남았다.

- EDA 또는 모델링에 미치는 영향:

날짜형 결측치는 고객의 가입 기간이나 최근 구매 시점과 같은 고객 행동 특성을 정확하게 분석하는 데 영향을 줄 수 있다.

또한 이상치가 포함된 상태에서 평균, 표준편차, 상관관계 등을 계산하면 일부 극단적인 값에 의해 전체적인 분포가 왜곡될 수 있으며, 회귀나 머신러닝 모델에서도 모델의 학습 결과에 영향을 줄 수 있다.

- 추가 확인이 필요한 조건 또는 한계:

이상치가 실제 고객의 정상적인 고액구매인지 입력 오류인지를 현재 데이터만으로는 명확하게 판단하기 어렵다. 

또한 날짜 결측치가 무작위로 발생한 것인지 특정 고객군에서 집중적으로 발생한 것인지 추가적으로 확인할 필요가 있다.

### 4-2. 결측치·이상치 처리 방법의 선택 근거

- 결측치 처리 방법:

수치형 변수는 변수의 특성과 분포를 고려하여 중앙값 또는 최빈값으로 결측치를 대체하고, 범주형 변수는 결측치를 unknown으로 처리하였다. 날짜형 변수의 경우 임의의 날짜를 입력하기보다 결측 상태(NaT)를 유지하는 방법을 적용하였다.

- 해당 방법을 선택한 이유:

수치형 변수에 평균을 사용할 경우 이상치의 영향을 받을 수 있기 때문에 상대적으로 극단값의 영향을 덜 받는 중앙값이나 최빈값을 활용하는 것이 적절하다고 판단하였다. 

범주형 변수는 임의의 값을 넣는 것(예: gender칼럼의 결측치에 Male, Female을 넣는 것보단)보단 결측치를 알수없음으로 표기해 따로 처리하는게 더 적합하다고 생각했다.

날짜형 변수는 임의의 값을 입력하면 실제 날짜로 오해될 수 있으므로, 날짜 정보가 없는 경우 NaT로 유지하는 것이 데이터의 의미를 보존하는 데 적절하다고 판단했다.

- 이상치 처리 방법:

수치형 변수의 이상치는 IQR(사분위범위)을 기준으로 하여 하한값과 상한값을 확인하고, 데이터의 특성을 고려하여 필요한 경우 범위를 벗어난 값을 조정하려고 하였으나, 

수치형 변수들을 보니 나이, 방문횟수, 표준/총 구매금액 모두 이상치가 실질적인 의미를 가진 데이터(예: 실제로 고령, 실제로 자주방문, 실제 고액구매)일수 있겠다고 판단하여 이상치 처리를 

시행하지는 않고 코드를 주석화 해놓았다. 

- 이상치 처리 시 주의점:

IQR 기준을 벗어났다고 해서 반드시 오류 데이터인 것은 아니다. 

특히 구매금액이나 방문횟수처럼 실제로 고객 간 차이가 크게 발생할 수 있는 변수에서는 극단값 자체가 중요한 고객 특성을 나타낼 수 있으므로, 단순히 이상치를 삭제하지 않고 실제 오류 여부와 분석 목적을 함께 고려해야 한다.

### 4-3. 파생변수의 활용 의도와 한계

- `AgeGroup` 생성 기준과 활용 목적:

Age를 연령 구간별로 구분하여 10대 이하, 10대, 20대, 30대, 40대, 50대, 60대 이상으로 그룹화하였다(데이터에 10대는 없었으므로). 

이를 통해 개별 연령을 사용하는 것보다 연령대별 고객 수, 구매금액, 방문횟수, 만족도 등의 차이를 비교하여 고객 특성을 파악할 수 있다.

- `PurchaseGrade` 생성 기준과 활용 목적:

TotalPurchaseAmount를 사분위수 기준으로 4개 그룹으로 나누어 Low, Medium, High, VIP로 분류하였다. 

이를 활용하면 구매금액 수준에 따른 고객 특성을 비교하고, 각 구매등급별 연령대, 방문횟수, 만족도, 선호 카테고리 등의 차이를 분석할 수 있다.

- 파생변수 사용 시 한계:

연속형 변수를 구간으로 나누면 원래 가지고 있던 세부적인 정보가 손실될 수 있다. 

예를 들어 같은 AgeGroup에 속한 고객이라도 실제 연령에는 차이가 있으며, PurchaseGrade 역시 동일한 등급 안에서 구매금액 차이가 존재할 수 있다. 

또한 PurchaseGrade는 현재 데이터의 사분위수를 기준으로 생성했기 때문에 데이터의 분포가 변경되면 각 등급의 기준도 달라질 수 있다. 

따라서 파생변수는 고객 특성을 쉽게 비교하기 위한 보조적인 변수로 활용하고, 필요에 따라 원래의 연속형 변수도 함께 확인하는 것이 필요할 수도 있다.
